pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeSegmentDataManager.java

```diff
@@ -475,7 +475,10 @@ protected boolean consumeLoop()
               messageBatch.getMessageCount(), messageBatch.getUnfilteredMessageCount(),
               messageBatch.isEndOfPartitionGroup());
         }
-        _endOfPartitionGroup = messageBatch.isEndOfPartitionGroup();
+        // We need to check for both endOfPartitionGroup and messageCount == 0, because
+        // endOfPartitionGroup can be true even if this is the last batch of messages (has been observed for kinesis)
+        // To process the last batch of messages, we need to set _endOfPartitionGroup to false in such a case
+        _endOfPartitionGroup = messageBatch.getMessageCount() == 0 && messageBatch.isEndOfPartitionGroup();
```

Comment:
When was this observed? IMO, we should fix the logic in KinesisMessageBatch that's resulting in this instead of making this change

Reply:
Got it. Then this makes sense. We can have the additional check for messages == 0 as well